File: typescript/Day.ts
export abstract class Day {
  year: number;
  day: number;

  abstract partOne(): string;
  abstract partTwo(): string;

  constructor(year: number, day: number) {
    this.year = year;
    this.day = day;
  }

  protected readInput(suffix?: string): string {
    const path = `../inputs/${this.year}_${
      this.day.toString().padStart(2, "0")
    }${suffix ? `_${suffix}` : ""}.txt`;

    return Deno.readTextFileSync(path);
  }
}

const DAY_REGISTRY: Map<string, new (year: number, day: number) => Day> =
  new Map();

export function getDay(year: number, day: number): Day | undefined {
  const DayConstructor = DAY_REGISTRY.get(`${year}-${day}`);
  if (DayConstructor == undefined) return undefined;

  return new DayConstructor(year, day);
}

export function RegisterDay(year: number, day: number) {
  return function <T extends new (year: number, day: number) => Day>(
    constructor: T,
  ) {
    DAY_REGISTRY.set(`${year}-${day}`, constructor);

    return constructor;
  };
}

File: csharp/Day.cs
using System.Reflection;

namespace aoc;

public abstract class Day
{
    public abstract string PartOne();
    public abstract string PartTwo();

    public int GetYear()
    {
        var attribute = this.GetType().GetCustomAttribute<RunnableDayAttribute>();
        return attribute?.Year ?? 0;
    }

    public int GetDay()
    {
        var attribute = this.GetType().GetCustomAttribute<RunnableDayAttribute>();
        return attribute?.Day ?? 0;
    }

    protected string[] ReadInputLines(string? name = null)
    {
        var fileName = GetFileName(name);
        var path = $"../inputs/{fileName}";

        if (!File.Exists(path))
        {
            throw new FileNotFoundException($"Input file not found: {path}");
        }

        return File.ReadAllLines(path);
    }

    protected string ReadInput(string? name = null)
    {
        return string.Join("", ReadInputLines(name));
    }

    private string GetFileName(string? name)
    {
        var year = GetYear();
        var day = GetDay();
        var dayPadded = day.ToString("D2");

        return name == null
            ? $"{year}_{dayPadded}.txt"
            : $"{year}_{dayPadded}_{name}.txt";
    }
}
